### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by jessy96**

### CODE REVIEW

File: src/components/Modal.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import "./Modal.css";
+import React from "react";
+
+export default function Modal({ handleClose, show, children }) {
+    const showHideModalClassName = show ? "modal display-block" : "modal display-none";
```
Comment (Vladyslav-Velytskyi) [inline src/components/Modal.js]: You can hide the modal by returning `null`

```
return !show ? null : (<div> ... </div>);
```

File: src/components/MovieCard.js [hunk 2]
```diff
@@ -1,19 +1,71 @@
...
-    <div className="movie-card">
-        <img src={require(`../static/images/movies/${movie.icon}`)}/>
-        <div className="movie-card-info">
-            <div className="movie-card-name">{movie.name}</div>
-            <div className="movie-card-genres">
-                {movie.genres.map(genre=>genre.name).join(", ")}
+const ContextMenu = React.lazy(()=> import("./ContextMenu"));
+
+
+const  MovieCard = function(props){
```
Comment (Vladyslav-Velytskyi) [inline src/components/MovieCard.js]: don't afraid to use arrow functions
```suggestion
const  MovieCard = (props) => {
```

File: src/components/MovieCard.js [hunk 3]
```diff
@@ -1,19 +1,71 @@
...
+    }
+    if(movie.year === undefined){
+        movie.year = 2000;
+    }
+    if(movie.icon === undefined){
+        movie.icon = "default.png";
+    }
+    if(movie.genres === undefined){
+        movie.genres = {}
+    }
```
Comment (Vladyslav-Velytskyi) [inline src/components/MovieCard.js]: it's a little bit hardcoded, you could try to add it into [`defaultProps`](https://ru.reactjs.org/docs/typechecking-with-proptypes.html)

File: src/components/MovieCard.js [hunk 4]
```diff
@@ -1,19 +1,71 @@
...
+                        {movie.genres.map(genre=>genre.name).join(", ")}
+                    </div>
+                    <div className="movie-card-year">{movie.year}</div>
+                </div>
             </div>
-            <div className="movie-card-year">{movie.year}</div>
-        </div>
-    </div>
-)
+            <Suspense fallback={<div className="error">Loading...</div>}>
```
Comment (Vladyslav-Velytskyi) [inline src/components/MovieCard.js:42]: Suspense is experimental component and I think that is not good idea to use it in educational project

File: src/components/UpsertMovieModal.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        console.log(contextValue);
+
+        contextValue.setMovieRepo(movieRepo.clone());
+        handleCloseModal();
+    };
+    const onError = (errors, e) => console.log(errors, e);
+
+    const handleUpdate = (movie, formData)=>{
+        if(formData.title && formData.title !==""){
```
Comment (Vladyslav-Velytskyi) [inline src/components/UpsertMovieModal.js]: you could just leave `formData.title`, in JS empty string in condition will convert to false

File: src/components/DeleteMovieModal.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import "./DeleteMovieModal.css";
+import React, { useContext } from "react";
+import MovieContext from "../services/MoviesContext";
+
+export default function DeleteMovieModal({handleClose, movie}){
+    const contextValue = useContext(MovieContext);
```
Comment (Vladyslav-Velytskyi) [inline src/components/DeleteMovieModal.js]: you could destructure this value

```suggestion
    const { movieRepo, setMovieRepo } = useContext(MovieContext);
    
    ...
    movieRepo.deleteMovie(movie);
    setMovieRepo(movieRepo.clone());
```

File: src/components/UpsertMovieModal.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
+import "./UpsertMovieModal.css";
+import React, {useContext} from "react";
+import { useForm } from "react-hook-form";
+import MoviesContext, { MovieRepo } from "../services/MoviesContext";
+import GenresRepo from "../services/GenreRepo";
+
+
+export default function UpsertMovieModal({header, movie, handleCloseModal}){
+    let contextValue = useContext(MoviesContext);
```
Comment (Vladyslav-Velytskyi) [inline src/components/UpsertMovieModal.js]:
```suggestion
    const contextValue = useContext(MoviesContext);
```

**Review by Vladyslav-Velytskyi — APPROVED** [review]

File: src/components/MovieCard.js (hunk 3 above)
Comment (jessy96) [reply src/components/MovieCard.js]: how do I use default props on nested object field level?

File: src/components/MovieCard.js (hunk 4 above)
Comment (jessy96) [reply src/components/MovieCard.js:42]: it was in the lecture, that why it is used - so I have it on record not to forget it if I need it in a future

File: src/components/Modal.js (hunk 1 above)
Comment (jessy96) [reply src/components/Modal.js]: done

File: src/components/MovieCard.js (hunk 2 above)
Comment (jessy96) [reply src/components/MovieCard.js]: done

File: src/components/UpsertMovieModal.js (hunk 5 above)
Comment (jessy96) [reply src/components/UpsertMovieModal.js]: done

File: src/components/DeleteMovieModal.js (hunk 6 above)
Comment (jessy96) [reply src/components/DeleteMovieModal.js]: done

File: src/components/UpsertMovieModal.js (hunk 7 above)
Comment (jessy96) [reply src/components/UpsertMovieModal.js]: done

**Review by Vladyslav-Velytskyi — APPROVED** [review]

**Pull request merged**